File: services/diseases-analyzers/app/services/rabbitmq_connection_service.py
import os
import pika
from dotenv import load_dotenv
import logging

load_dotenv()

class RabbitMQConnectionService:
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.rabbitmq_host = os.getenv('RABBIT_MQ_HOST')
        self.rabbitmq_port = os.getenv('RABBIT_MQ_PORT')
        self.rabbitmq_user = os.getenv('RABBIT_MQ_DEFAULT_USER')
        self.rabbitmq_password = os.getenv('RABBIT_MQ_DEFAULT_PASS')
        self.connection = None
        self.channel = None


    def set_up_connection(self):
        self.connection = pika.BlockingConnection(
            pika.ConnectionParameters(
                host=self.rabbitmq_host,
                port=self.rabbitmq_port,
                credentials=pika.PlainCredentials(
                    self.rabbitmq_user,
                    self.rabbitmq_password
                )
            )
        )
        self.channel = self.connection.channel()

    def assert_queue(self, queue: str):
        self.logger.debug(f"Asserting queue - {queue}")
        self.channel.queue_declare(queue=queue, durable=True)

    def get_channel(self):
        return self.channel

File: services/api/src/systems/message-queue/rabbit-mq-connection.service.ts
import { Injectable, Logger, OnModuleInit } from '@nestjs/common';
import * as amqp from 'amqplib';
import { AppConfigService } from 'src/config/app-config.service';
import { DiagnosticModelsService } from 'src/modules/diagnostic-models/diagnostic-models.service';

@Injectable()
export class RabbitMQConnectionService implements OnModuleInit {
  private readonly logger = new Logger(RabbitMQConnectionService.name);

  private readonly rabbitMQProtocol =
    this.configService.get<string>('RABBIT_MQ_PROTOCOL');
  private readonly rabbitMQHost =
    this.configService.get<string>('RABBIT_MQ_HOST');
  private readonly rabbitMQUser = this.configService.get<string>(
    'RABBITMQ_DEFAULT_USER',
  );
  private readonly rabbitMQPassword = this.configService.get<string>(
    'RABBITMQ_DEFAULT_PASS',
  );
  private connection: amqp.Connection;
  private channel: amqp.Channel;

  constructor(
    private readonly configService: AppConfigService,
    private readonly diagnosticModelsService: DiagnosticModelsService,
  ) {}

  async onModuleInit() {
    await this.setUpConnection();
  }

  async setUpConnection() {
    this.connection = await amqp.connect(
      `${this.rabbitMQProtocol}://${this.rabbitMQUser}:${this.rabbitMQPassword}@${this.rabbitMQHost}`,
    );
    this.channel = await this.connection.createChannel();
    await this.assertQueueForModels();
  }

  async assertQueue(queue: string) {
    this.logger.debug(`Asserting queue - ${queue}`);
    await this.channel.assertQueue(queue, { durable: true });
  }

  async assertQueueForModels() {
    const modelQueueNames =
      await this.diagnosticModelsService.selectModelQueueNames();

    for (const modelQueueName of modelQueueNames) {
      await this.assertQueue(modelQueueName);
    }
  }

  getChannel() {
    return this.channel;
  }
}
